### DISCUSSION TITLE
Support for Sphinx>=4.0

### DISCUSSION THREAD

**Discussion by hoefling** [pradyunsg/furo#148] [Potential Issue] [answered]
Right now, `requires = ["sphinx ~= 3.0"]` prevents from switching to Sphinx==4.0. Requesting broadening of the dependency range. I tested the version bump locally and see no degradation in build logs (no new warnings for 4.0) or kitchen sink. @pradyunsg what do you think?

**Comment by zerocewl** [comment]
+1

**Comment by pradyunsg** [comment, marked as answer]
Sounds reasonable to me! I'll need to test locally, to make sure that things all work right.

**Reply by hoefling** [reply]
@pradyunsg thank you! If you need additional input, feel free to ping me; I have tried the bump out on our (quite extensive) internal docs, everything looks good so far with `sphinx==4.0.1` and `furo==2021.4.11b34` (latest releases).

**Reply by pradyunsg** [reply]
Done in beta 35! :)

**Reply by hoefling** [reply]
@pradyunsg awesome, thank you! :rocket:

**Reply by pradyunsg** [reply]
WELL, there was one thing broken by the upgrade that I missed in the initial release: syntax highlighting in dark mode -- e49e5dde09a71998dc334929921083e8b071fc6c

**Comment by vberlier** [comment]
Any update? I love furo but right now it's the only thing keeping me from updating to Sphinx 4.

**Reply by pradyunsg** [reply]
Appreciate the nudge! :)

**Comment by mcepl** [comment]
Is this related to Sphinx 4.1.1 being used?
```
petunia_dursley@kusansky (master *+)$ make html
Running Sphinx v4.1.1
loading translations [en_GB]... not available for built-in messages
loading pickled environment... done

Theme error:
no theme named 'furo' found (missing theme.conf?)
make: *** [Makefile:20: html] Chyba 2
petunia_dursley@kusansky (master *+)$ 
```

and everything seems to be installed correctly:
```
petunia_dursley@kusansky (master *+)$ ls /usr/lib/python3.8/site-packages/furo/theme/furo/theme.conf
/usr/lib/python3.8/site-packages/furo/theme/furo/theme.conf
petunia_dursley@kusansky (master *+)$ ipython
Python 3.8.10 (default, May 05 2021, 15:36:36) [GCC]
Type 'copyright', 'credits' or 'license' for more information
IPython 7.25.0 -- An enhanced Interactive Python. Type '?' for help.

In [1]: import furo

In [2]: ??furo.setup
Signature: furo.setup(app: sphinx.application.Sphinx) -> Dict[str, Any]
Source:
def setup(app: sphinx.application.Sphinx) -> Dict[str, Any]:
    """Entry point for sphinx theming."""
    app.require_sphinx("3.0")

    app.add_config_value(
        "pygments_dark_style", default="native", rebuild="env", types=[str]
    )

    app.add_html_theme("furo", str(THEME_PATH))

    app.connect("html-page-context", _html_page_context)
    app.connect("builder-inited", _builder_inited)
    app.connect("build-finished", _overwrite_pygments_css)

    return {
        "parallel_read_safe": True,
        "parallel_write_safe": True,
        "version": __version__,
    }
File:      /usr/lib/python3.8/site-packages/furo/__init__.py
Type:      function

In [3]:
Do you really want to exit ([y]/n)?
petunia_dursley@kusansky (master *+)$
```

(works perfectly well with `alabaster` theme).

**Reply by pradyunsg** [reply]
It's likely your environment-- check if python --version and `where python` matches what the same commands with ipython will output.

**Reply by mcepl** [reply]
> It's likely your environment-- check if python --version and `where python` matches what the same commands with ipython will output.

You are kidding, aren’t you? I am the lead packager of Python at SUSE, of course, I know what my Pythons are about, and no, I don't use any other Python than the system one:
```
petunia_dursley@kusansky (master *)$ python3 --version
Python 3.8.11
petunia_dursley@kusansky (master *)$ ipython3
Python 3.8.11 (default, Aug 10 2021, 00:29:46) [GCC]
Type 'copyright', 'credits' or 'license' for more information
IPython 7.25.0 -- An enhanced Interactive Python. Type '?' for help.

In [1]: import furo

In [2]:
Do you really want to exit ([y]/n)?
petunia_dursley@kusansky (master *)$ make html SPHINXOPTS="-v"
Running Sphinx v4.1.2
loading translations [en_GB]... locale_dir /home/matej/archiv/2016/clanky/petunia_dursley/locales/en_GB/LC_MESSAGES does not exists
locale_dir /home/matej/archiv/2016/clanky/petunia_dursley/locales/en_GB/LC_MESSAGES does not exists
not available for built-in messages
loading pickled environment... done

Traceback (most recent call last):
  File "/usr/lib/python3.8/site-packages/sphinx/cmd/build.py", line 276, in build_main
    app = Sphinx(args.sourcedir, args.confdir, args.outputdir,
  File "/usr/lib/python3.8/site-packages/sphinx/application.py", line 270, in __init__
    self._init_builder()
  File "/usr/lib/python3.8/site-packages/sphinx/application.py", line 327, in _init_builder
    self.builder.init()
  File "/usr/lib/python3.8/site-packages/sphinx/builders/html/__init__.py", line 219, in init
    self.init_templates()
  File "/usr/lib/python3.8/site-packages/sphinx/builders/html/__init__.py", line 267, in init_templates
    self.theme = theme_factory.create(themename)
  File "/usr/lib/python3.8/site-packages/sphinx/theming.py", line 245, in create
    raise ThemeError(__('no theme named %r found (missing theme.conf?)') % name)
sphinx.errors.ThemeError: no theme named 'furo' found (missing theme.conf?)

Theme error:
no theme named 'furo' found (missing theme.conf?)
make: *** [Makefile:20: html] Chyba 2
petunia_dursley@kusansky (master *)$
```

Besides, how come just when I change the value of variable `html_theme` in `conf.py` from `"furo"` to `"alabaster"` everything works?

**Reply by pradyunsg** [reply]
Please file an separate discussion for this, with clear steps on how to reproduce this independently. Please include the list of packages installed using `python3 -m pip list`, and invoke sphinx using `python3 -m sphinx` rather than `make html` in that reproducer.

**Comment by josuemtzmo** [comment]
I love the theme, thanks for the great work! I was having the same issue using `Sphinx 3.5.4`, after updating to `Sphinx 4.0.2` the issue was solved.

**Comment by mcepl** [comment]
> Please file an separate discussion for this, with clear steps on how to reproduce this independently. Please include the list of packages installed using `python3 -m pip list`, and invoke sphinx using `python3 -m sphinx` rather than `make html` in that reproducer.

https://github.com/pradyunsg/furo/issues/231

Does the Code of Conduct bans me from thinking that this form-driven new system is intended to get rid of as many reporters as possible?

**Reply by pradyunsg** [reply]
> Does the Code of Conduct bans me from thinking that this form-driven new system is intended to get rid of as many reporters as possible?

I'm not sure how that is related to the CoC. The Code of Conduct is designed to prevent harassment in this project's spaces, and unrelated to whatever you might think about the project.

The issue forms are so that issues filed provide clear context for what needs to be done, to me. I don't expect most users to file issues. The intent is to push users who are looking for support to open discussions in this forum instead, so that I don't have to do user support over GitHub Issues.

For some reason, despite me having asked you to open a GitHub discussion, you've gone ahead and filed an issue instead. That was your prorogative. It's fine. I don't think we should spam OP and others for the problems you're facing and we can conduct the tech support for you there -- assuming you don't put me off from spending my free time helping you. :)
